src/test/scala/mesosphere/marathon/integration/setup/ServiceMock.scala

```diff
@@ -0,0 +1,199 @@
+      response.setHeader("Content-Type", "application/json")
+      response.setStatus(code)
+    }
+    def ok[T](t: T)(implicit writes: Writes[T]): Unit = status(t, 200)
+    def error[T](t: T)(implicit writes: Writes[T]): Unit = status(t, 503)
+
+    (request.getMethod, request.getPathInfo) match {
+      case ("GET", "/v1/plan") =>
+        if (plan.isDone) ok(plan) else error(plan)
+      case ("PUT", "/v1/plan") =>
+        val cmd = Option(request.getParameter("cmd")).getOrElse("undefined")
+        cmd match {
+          case "continue" => plan.next(withContinue = true)
+          case "finalize" => plan.doFinalize()
+        }
+        ok(Json.obj("Result" -> s"Received cmd: $cmd"))
+      case ("POST", "/admin/rollback") =>
+        plan.rollback()
+        ok(Json.obj("Result" -> "Rollback"))
+      case ("PUT", "/admin/next") =>
+        plan.next(withContinue = false)
+        ok(Json.obj("Result" -> plan.status))
+      case ("POST", "/admin/error") =>
+        plan.errors ::= s"Got an error at ${new Date}"
+        ok(plan.errors)
+    }
+    baseRequest.setHandled(true)
+  }
+}
+
+object ServiceMock {
+
+  case class Plan(phases: List[Phase]) {
+    var errors: List[String] = List.empty
+    def isDone = phases.forall(_.isDone)
+    def status = currentPhase.fold("Complete")(_.status)
+    def currentPhase: Option[Phase] = phases.find(_.currentBlock.isDefined)
+    def next(withContinue: Boolean): Boolean = {
+      currentPhase.fold(true) { phase =>
+        val done = phase.next(withContinue)
+        if (done) currentPhase.map(_.next(withContinue)).getOrElse(true) else done
+      }
+    }
+    def doFinalize(): Unit = {
+      phases.foreach(_.blocks.foreach(_.doFinalize()))
+    }
+    def rollback(): Unit = {
+      phases.foreach(_.rollback)
+      errors = List.empty
+    }
+  }
+
+  case class Phase(name: String, blocks: List[Block]) {
+    def isDone = blocks.forall(_.isDone)
+    def status = currentBlock.fold("Complete")(_.status)
+    def currentBlock: Option[Block] = blocks.find(b => b.isInProgress || b.notStarted)
+    def next(withContinue: Boolean): Boolean = {
+      currentBlock.fold(true) { block =>
+        val done = block.next(withContinue)
+        if (done) block.next(withContinue) else done
+      }
+    }
+    def rollback(): Unit = blocks.foreach(_.rollback)
+  }
+
+  case class Block(name: String, decisionPoint: Boolean) {
+    private var inProgress: Boolean = false
+    private var done: Boolean = false
+
+    def isDone = done
+    def isInProgress = inProgress
+    def notStarted = !inProgress && !done
+
+    def status: String = {
+      if (done) "Complete"
+      else if (inProgress && decisionPoint) "Waiting"
+      else if (inProgress && !decisionPoint) "InProgress"
+      else "Pending"
+    }
+
+    def next(withContinue: Boolean): Boolean = {
+      if (notStarted) inProgress = true
+      else if (inProgress && ((decisionPoint && withContinue) || (!decisionPoint && !withContinue))) {
```

Comment:
I am not sure if that's important but your model is missing the "InProgress" state after the "Waiting" state. It switches directly to "Complete"